File: ex14.py

def divisors(n):
    assert n > 0
    for t in range(1, int(n**0.5)+1):
        if n % t == 0:
            yield t
            if (d := n // t) != t:
                yield d

def is_prime(n):
    return len(tuple(divisors(n))) == 2

def is_perfect(n):
    return sum(divisors(n)) == 2*n

def perfects(max):
    return (n for n in range(1, max+1) if is_perfect(n))


if __name__ == "__main__":
    print(list(divisors(6)))
    print(list(divisors(1024)))
    print(is_prime(7))
    print(is_prime(24))
    print(is_prime(1))
    print(tuple(perfects(10000)))

File: ex14.rs

fn divisors(n:u64) -> Vec<u64> {
    assert!(n != 0);
    let mut r: Vec<u64> = Vec::new();
    for t in 1..=((n as f64).sqrt().floor() as u64) {
        if n % t == 0 {
            r.push(t);
            let d = (n / t) as u64;
            if d != t {
                r.push(d);   
            }
        }
    }
    r
}

fn is_perfect(n: u64) -> bool {
    divisors(n).iter().sum::<u64>() == 2*n
}

fn perfects(max: u64) -> Vec<u64> {
    let mut r: Vec<u64> = Vec::new();
    for n in 1..=max {
        if is_perfect(n) {
            r.push(n);
        }
    }
    r
}

fn main() {
    println!("div 6 : {:?}", divisors(6));
    println!("div 1024 : {:?}", divisors(1024));
    println!("perfect 6 : {:?}", is_perfect(6));
    println!("perfect 10 : {:?}", is_perfect(10));
    println!("perfects 1000000 : {:?}", perfects(1000000));
}